src/Telemetry.h

```diff
@@ -0,0 +1,100 @@
+ * a stream, but the entire implementation can be replaced to log in arbitrary ways.
+ */
+
+#include <iostream>
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "Expr.h"
+#include "Target.h"
+
+namespace Halide {
+namespace Internal {
+
+class Telemetry {
+public:
+    Telemetry() = default;
+    virtual ~Telemetry() = default;
+
+    /** Record when a particular simplifier rule matches.
+     */
+    virtual void record_matched_simplifier_rule(const std::string &rulename) = 0;
+
+    /** Record when an expression is non-monotonic in a loop variable.
+     */
+    virtual void record_non_monotonic_loop_var(const std::string &loop_var, Expr expr) = 0;
+
+    /** Record when can_prove() fails, but cannot find a counterexample.
+     */
+    virtual void record_failed_to_prove(Expr failed_to_prove, Expr original_expr) = 0;
+
+    /**
+     * Emit all the gathered data to the given stream. This may be called multiple times.
+     */
+    virtual std::ostream &emit_to_stream(std::ostream &o) = 0;
+};
+
+/** Set the active Telemetry object, replacing any existing one.
+ * It is legal to pass in a nullptr (which means "don't do any telemetry").
+ * Returns the previous Telemetry (if any). */
+std::unique_ptr<Telemetry> set_telemetry(std::unique_ptr<Telemetry> telemetry);
+
+/** Return the currently active Telemetry object. If set_telemetry()
+ * has never been called, a nullptr implementation will be returned.
+ * Do not save the pointer returned! It is intended to be used for immediate
+ * calls only. */
+Telemetry *get_telemetry();
+
+/** JSONTelemetry is a basic implementation of the Telemetry interface
+ * that saves logged data, then logs it all in JSON format in emit_to_stream().
+ */
+class JSONTelemetry : public Telemetry {
+public:
+    JSONTelemetry() = default;
+
+    JSONTelemetry(
+        const std::string &generator_name,
+        const std::string &function_name,
+        const std::string &autoscheduler_name,
+        const Target &target,
+        const std::string &generator_args,
+        bool anonymize_exprs);
+
+    void record_matched_simplifier_rule(const std::string &rulename) override;
+    void record_non_monotonic_loop_var(const std::string &loop_var, Expr expr) override;
+    void record_failed_to_prove(Expr failed_to_prove, Expr original_expr) override;
+    std::ostream &emit_to_stream(std::ostream &o) override;
+
+protected:
+    const std::string generator_name;
+    const std::string function_name;
+    const std::string autoscheduler_name;
+    const Target target;
+    const std::string generator_args;
+    const bool anonymize_exprs{false};
+
+    std::map<std::string, int64_t> matched_simplifier_rules;
+    // TODO: are we likely to see enough duplicates here to make
+    // it worth considering a set instead of a vector?
+    std::vector<std::pair<std::string, Expr>> non_monotonic_loop_vars;
```

Comment:
Yeah, the same ones will be encountered multiple times during lowering.

Reply:
will fix then